docs_src/design/adr/device-service/0021-Profile-Changes.md

```diff
@@ -0,0 +1,101 @@
+| Device Commands | array | Situational | When the device profile is not associated to any device, reading or event, the device commands list can be modified (i.e., adding or removing the associated device resources).  Once the profile is associated to a device, reading or event, only new commands can be added to the array.  Deleting is prohibited.  Device commands are optional and so the array can always be empty. |
+| *Device Resource Properties* |||
+| Description | string | Anytime | As this property does not affect any functionality, this property can be change at any time – to include setting the string to an empty string. |
+| Name | string | Situational | When the device profile is not associated to any device, reading or event, this field can be modified.  It can never be blank and must be unique across the entire profile. |
+| IsHidden | bool | Anytime | This field allows the device resource to be exposed (default) via the command service.  This change does not affect the behavior but only visibility.  This field is optional but is false by default when not provided. |
+| Tag | string | Situational | As this property does not affect any functionality, this property can be change at any time – to include setting the string to an empty string. |
+| Properties | Resource Properties | Situational | When the device profile is not associated to any device, reading or event, this field can be modified.  Because the ValueType and ReadWrite properties of this field are required, it can never be empty. |
+| Attributes | map | Situational | When the device profile is not associated to any device, reading or event, this field can be modified.|
+| *Device Command Properties* ||||
+| Name | string | Situational | When the device profile is not associated to any device, reading or event, this field can be modified.  It can never be blank and must be unique across the entire profile. |
+| IsHidden | bool | Anytime | This field allows the device command to be exposed (default) via the command service.  This change does not affect the behavior but only visibility.  This field is optional but is false by default when not provided.|
+| ReadWrite | string | Situational | When the device profile is not associated to any device, reading or event, this field can be modified.  This field is required.  As the field is expected to be from an enumeration of R, W and RW, it can never be empty. |
+| ResourceOperations | array | Situational | When the device profile is not associated to any device, reading or event, this field can be modified.  This field is required and can never be empty. |
+
+ 
+### Empty Profile
+
+A device profile can begin life “empty” - meaning that it has no device resources or device commands, but only when the profile is not associated to any device, event or reading.  As soon as it is associated to any other EdgeX object, it must have at least one device resource.  Device commands are always optional.
```

Comment:
why can't users add a Device with an empty Device Profile?

Reply:
I was informed (can't remember by who) that a profile with an empty set of device resources, etc. would create problems in the device service.  However, this is current state as I understand it.  So if we wish to allow devices with profiles and no device resources and provide the appropriate checks to allow this to work, I am ok with it.

Question might be - why?  Why would anyone want a device without device resources.  An empty device profile serves as a place holder, but an actual device without device resources??? Would the device just be a placeholder too then?

Reply:
I added a section to cover the "summary" statements made by Cloud above

Reply:
I don't see my summary is added to the ADR.  It looks like we just block the profile changes when it is associated to something.  Why don't we add those new API?  If we only support to update the whole profile (PUT API), why do we create the table in this ADR?
If the empty profile cannot be used, why do they need a placeholder?  If we have those Device Resource and Device Command creation API, an empty profile can be used by a Device.  Then the device resources and device commands can be added in the future.

Reply:
_I don't see my summary is added to the ADR._ 
  Added your proposed APIs in **proposed design section**

_It looks like we just block the profile changes when it is associated to something._ 
correct - as stated in the summary

_Why don't we add those new API? If we only support to update the whole profile (PUT API), why do we create the table in this ADR?_
So you are not allowing a change to an attribute or a property (for example)?  Rather you are only allowing for the add/delete of a device resoruces or device commands?  Why not offer a PATCH to these fields?  I added a to be determined note in the API proposed design section to raise this.

The table helps to make clear the allowed changes (independent of any API).  APIs should be written to codify what's in the table (once it is agreed upon) - regardless of which API or HTTP operation.  

_If the empty profile cannot be used, why do they need a placeholder?_ 
It can be used.  We said in the table that the profile's resources and commands can be modified until it is associated to an object.  At which point, then new resources or commands can be added.  I made this change to the document.

_If we have those Device Resource and Device Command creation API, an empty profile can be used by a Device. Then the device resources and device commands can be added in the future._
Agree - per above and as stipulated in the table.  I added more text around this to make it clear.